### PR TITLE
feat!: ecocredit genesis import and export

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #187 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:15]:
```suggestion
  // class_info is the list of credit class info.
```
information has no plural form, ditto for the fields/comments below

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:56]: why not just using `address` instead (as in x/bank genesis)?
```suggestion
  // address is the account address of the account holding credits.
  string address = 1;
```

**Review by blushi — COMMENTED** [review]
I haven't looked at the implementation yet, just some comments regarding the proto definitions

File: x/ecocredit/server/testsuite/suite.go
Comment (ruhatch) [inline x/ecocredit/server/testsuite/suite.go:69]: I guess we can also remove the `ctx` parameter, as that is using `s.sdkCtx`?

**Review by ruhatch — APPROVED** [review]
This looks good to me. We might want some more eyes on it though, as there are a lot of bits and I'm not very experienced with Go yet!

File: x/ecocredit/server/genesis.go
Comment (blushi) [inline x/ecocredit/server/genesis.go:90]: strictly speaking, there's no keeper in x/ecocredit so maybe we should upgrade the language here

**Review by blushi — COMMENTED** [review]
Looks good overall, just some refactoring requests

**Review by robert-zaremba — COMMENTED** [review]
Proper go doc comments are missing

**Comment by blushi** [comment]
> Proper go doc comments are missing

@aleem1314 could you address that? I believe it's mainly in data_prefixes.go

File: x/ecocredit/go.mod
Comment (aaronc) [inline x/ecocredit/go.mod:14]: Why is this needed?

Comment (aleem1314) [reply x/ecocredit/go.mod:14]: Not using anywhere, removing it.

Comment (ruhatch) [reply x/ecocredit/go.mod]: This was from the retirement location stuff, but should have been removed - perhaps this snuck in a merge/rebase at some point, then disappeared after squash/merge.

File: x/ecocredit/server/genesis.go
Comment (aaronc) [inline x/ecocredit/server/genesis.go:64]: Supplies should only be used as an invariant check. Setting balances should update the supply and then we should check that the supply is what is expected. We should never just set supply based on a value in genesis.

**Review by aaronc — CHANGES_REQUESTED** [review]

File: x/ecocredit/server/invariants.go
Comment (anilCSE) [inline x/ecocredit/server/invariants.go:82]: Don't we  need to set `broken` here?

**Review by anilCSE — COMMENTED** [review]
lgtm, just one question

**Review by blushi — COMMENTED** [review]
Could you add some tests for the invariants? (as done in x/group)

**Review by blushi — APPROVED** [review]
LGTM!

**Comment by ruhatch** [comment]
@aleem1314 looks like you need to `make proto-gen` and then this is good to go :)

**Review by atheeshp — APPROVED** [review]
lgtm

**Comment by blushi** [comment]
@aleem1314 is this good to merge?

**Comment by aleem1314** [comment]
@aaronc could you have a final look?

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:34]:
```suggestion
  repeated Balance balances = 5;

  // tradable_supplies is the list of credit batch tradable supply.
  repeated Supply supplies = 6;

```

**Review by aaronc — APPROVED** [review]
Logic looks correct.

Can we simplify this API? Won't block on it because I know this has been in progress for a while, so pre-approving. But I think we can merge tradable and retired balances and supplies.

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (ruhatch) [reply proto/regen/ecocredit/v1alpha1/genesis.proto:34]: Would this mean adding an enum to the `Balance` type to keep track of if it's retired or tradable?

Comment (aleem1314) [reply proto/regen/ecocredit/v1alpha1/genesis.proto]: updated

Comment (clevinson) [reply proto/regen/ecocredit/v1alpha1/genesis.proto]: I just spoke with @aaronc about this and actually the intention of his comment was to still have separate fields for `retired_balance` and `tradable_balance`, but not have them be entirely separate `Balance` / `Supply` messages (for tradable / retired).

I think enums over complicate things, and we don't want/need to be tracking any balances outside of those that are retired or tradable (e.g. cancelled credits should never be tracked in genesis import/export).

So this would look something more like:
```proto
message Balance {
   // address is the account address of the account holding credits.
   string address = 1;

   // batch_denom is the unique ID of the credit batch.
   string batch_denom = 2;

   // tradable_balance is the tradable balance of the credit batch.
   string tradable_balance = 3;

   // retired_balance is the retired balance of the credit batch.
   string retired_balance = 3;
 }
 ```

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (clevinson) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:71]: I think adding an enum actually overcomplicates things.  @aleem1314 can we change this to something like:
```
message Balance {
   // address is the account address of the account holding credits.
   string address = 1;

   // batch_denom is the unique ID of the credit batch.
   string batch_denom = 2;

   // tradable_balance is the tradable balance of the credit batch.
   string tradable_balance = 3;

   // retired_balance is the retired balance of the credit batch.
   string retired_balance = 3;
 }
 ```
 
 I spoke with @aaronc today and confirmed that this was his original idea as well when he suggested combining into a singular `Balance` message.

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (clevinson) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:80]: Can you update this to the same format as above (e.g. with separate `tradable_balance` and `retired_balance` fields in the `Supply` message?)

**Review by clevinson — CHANGES_REQUESTED** [review]
Few comments on the formatting of the Supply / Balance messages. Can you take a look at this tomorrow @aleem1314 and then I think we're good to merge!

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (robert-zaremba) [reply proto/regen/ecocredit/v1alpha1/genesis.proto]: retired balance ID  should be `4`,  right?

File: proto/regen/ecocredit/v1alpha1/genesis.proto
Comment (robert-zaremba) [inline proto/regen/ecocredit/v1alpha1/genesis.proto:7]: I was posting similar comment to NFT: how about putting genesis into a separate package, eg `x/ecocredit/genesis` - it has different domain than RPC messages. Or even we can have two packages:
* x/ecocredit/  -- for all things related to RPC
* x/ecocredit/DAL -- for all things related to storage (data access layer)

Comment (clevinson) [reply proto/regen/ecocredit/v1alpha1/genesis.proto:7]: I agree with this separation in principle, but don't want to block this PR on it. adding a new versioning scheme outside of the one for just modules (which is the current setup) is something that I'd like to make sure is the direction we want to go in across the SDK before doing it just on our own modules.

We will already need to be doing a version upgrade for ecocredit module  proto files in regen ledger v3 as we expect the API to change possibly quite a bit. So seems fine IMO to not block the v2.0 upgrade on this question.

**Review by clevinson — APPROVED** [review]
Few merge conflicts now, but pending those getting fixed, I think this is good to go.

There's some stuff where we may want to add some additional validation to `InitGenesis`, namely that all the `CreditType` values in any given genesis credit classes actually correspond to a CreditType in the ecocredit's genesis params.

Happy to move that to a separate issue and just continue a discussion there to see what makes the most sense after this PR gets merged.

Thanks @aleem1314!

**Pull request merged**
